File: packages/effect-lightning/src/lightning-renderer.ts
/**
 * Lightning Canvas2D PANEL drawing (web).
 *
 * PERFORMANCE: the original web lightning was a single full-screen fragment pass
 * that, at EVERY pixel, walked the main bolt + every fork (8 × 14 segments) and
 * evaluated `boltPoint` — which calls 4-octave `fbm` TWICE — per segment: ~220
 * fbm (~3.5K hash) evaluations PER PIXEL, plus the shadow pass re-walking it 9×.
 * Fine on a GPU; ~1.1 s/frame under software/ANGLE WebGL. The bolt is a thin
 * polyline whose vertices are fragment-INDEPENDENT, so it belongs in a panel: we
 * compute the jagged polyline ONCE per frame in JS (a faithful port of the
 * shader's fbm/hash + boltPoint) and stroke it — soft halo into R, hot core into
 * G — into an offscreen Canvas2D. The fragment shader then just samples that and
 * adds the (cheap, full-screen) flash + impact glow + finish.
 *
 * Swift/Metal lightning is untouched; lightning.dope.json is unchanged across
 * platforms — only the web render path moved.
 *
 * Panel channel encoding consumed by lightning-shader.ts:
 *   R = soft electric HALO (glow)   ·   G = hot white CORE
 */

import { MAX_FORKS, BOLT_SEGS } from "./lightning-shader.js";
import { strikeProgress } from "./lightning-tempo.js";

/** Resolved render params the lightning panel consumes. */
export interface LightningRenderParams {
  durationMs: number;
  style: number;       // = whimsy (photoreal plasma 0 -> cel comic bolt 1)
  thickness: number;   // bolt half-width as fraction of min dim
  jagged: number;      // fbm perturbation amount of the bolt vertices
  branches: number;    // number of secondary forks
  boltSeed: number;    // per-fire hash offset
}

const clamp = (x: number, lo: number, hi: number): number => (x < lo ? lo : x > hi ? hi : x);
const mix = (a: number, b: number, t: number): number => a + (b - a) * t;
const fract = (x: number): number => x - Math.floor(x);

// --- Faithful JS port of the shared look/glsl hash + value-noise fbm ---------
function hash11(p: number): number {
  p = fract(p * 0.1031);
  p *= p + 33.33;
  p *= p + p;
  return fract(p);
}
function hash21(p: number): { x: number; y: number } {
  let x = fract(p * 0.1031), y = fract(p * 0.103), z = fract(p * 0.0973);
  const d = x * (y + 33.33) + y * (z + 33.33) + z * (x + 33.33);
  x += d; y += d; z += d;
  return { x: fract((x + y) * z), y: fract((x + z) * y) };
}
function vnoise(x: number, y: number): number {
  const ix = Math.floor(x), iy = Math.floor(y);
  const fx = x - ix, fy = y - iy;
  const ux = fx * fx * (3 - 2 * fx), uy = fy * fy * (3 - 2 * fy);
  const a = hash11(ix * 1 + iy * 57);
  const b = hash11((ix + 1) * 1 + iy * 57);
  const c = hash11(ix * 1 + (iy + 1) * 57);
  const d = hash11((ix + 1) * 1 + (iy + 1) * 57);
  return mix(mix(a, b, ux), mix(c, d, ux), uy);
}
function fbm(x: number, y: number): number {
  let s = 0, a = 0.5;
  for (let i = 0; i < 4; i++) {
    s += a * vnoise(x, y);
    const nx = (0.8 * x + 0.6 * y) * 2.03;
    const ny = (-0.6 * x + 0.8 * y) * 2.03;
    x = nx; y = ny; a *= 0.5;
  }
  return s;
}

interface Vec2 { x: number; y: number }

/** A jagged bolt vertex at parameter t along A→B (port of the shader boltPoint). */
function boltPoint(
  A: Vec2, B: Vec2, t: number, seedOff: number, jitterScale: number,
  seed: number, jagged: number, beat: number,
): Vec2 {
  const dx = B.x - A.x, dy = B.y - A.y;
  const len = Math.max(Math.hypot(dx, dy), 1);
  const dirx = dx / len, diry = dy / len;
  const nrmx = -diry, nrmy = dirx;
  const bt = beat * jitterScale;
  const n = fbm(t * 6 + seedOff + seed, bt * 0.5) - 0.5;
  const fine = fbm(t * 22 + seedOff * 3.1 + seed, bt) - 0.5;
  const taper = Math.sin(t * Math.PI);
  const off = (n * 1.6 + fine * 0.5) * jagged * len * 0.16 * taper;
  return { x: A.x + dirx * (t * len) + nrmx * off, y: A.y + diry * (t * len) + nrmy * off };
}

/** Build the drawn portion (0..`drawn`) of a jagged polyline A→B. */
function boltPolyline(
  A: Vec2, B: Vec2, drawn: number, seedOff: number, jitterScale: number,
  seed: number, jagged: number, beat: number,
): Vec2[] {
  const pts: Vec2[] = [boltPoint(A, B, 0, seedOff, jitterScale, seed, jagged, beat)];
  for (let i = 1; i <= BOLT_SEGS; i++) {
    const t = i / BOLT_SEGS;
    if (t - 1 / BOLT_SEGS > drawn) break;
    const tc = Math.min(t, drawn);
    pts.push(boltPoint(A, B, tc, seedOff, jitterScale, seed, jagged, beat));
  }
  return pts;
}

/**
 * Stroke a polyline as a tight soft HALO (red channel) + hot CORE (green),
 * additively. The halo is a thin bright stroke with a `shadowBlur` gaussian
 * falloff — a real bright-spine→transparent-edge glow, NOT a wide flat band (the
 * latter reads as a fuzzy translucent slab once the screen blend + gain amplify
 * its low-alpha tail).
 */
function strokeBolt(ctx: CanvasRenderingContext2D, pts: Vec2[], rad: number): void {
  if (pts.length < 2) return;
  ctx.beginPath();
  ctx.moveTo(pts[0].x, pts[0].y);
  for (let i = 1; i < pts.length; i++) ctx.lineTo(pts[i].x, pts[i].y);
  ctx.lineCap = "round";
  ctx.lineJoin = "round";

  // HALO -> R: a few TIGHT widening strokes approximate the plasma glow falloff
  // (bright spine, quick fade). Kept narrow (≤~1.6× rad) so it reads as a halo,
  // not a slab — and cheap (no shadowBlur gaussian, which is costly per frame).
  const halo: Array<[number, number]> = [[1.6, 0.16], [1.0, 0.3], [0.55, 0.6]];
  for (const [wmul, alpha] of halo) {
    ctx.lineWidth = Math.max(rad * wmul, 1);
    ctx.strokeStyle = `rgba(255,0,0,${alpha})`;
    ctx.stroke();
  }

  // CORE -> G: a crisp thin white-hot centre line.
  ctx.strokeStyle = "rgba(0,255,0,1)";
  ctx.lineWidth = Math.max(rad * 0.28, 1.5);
  ctx.stroke();
}

/**
 * Draw one frame of the bolt (main trunk + forks) into the offscreen panel.
 * `center` is the strike point (anchor, device px); the bolt descends from the
 * top edge to it. Geometry is a pure function of elapsedMs + seed.
 */
export function drawLightningPanel(
  ctx: CanvasRenderingContext2D,
  w: number,
  h: number,
  params: LightningRenderParams,
  elapsedMs: number,
  center: { x: number; y: number },
): void {
  ctx.clearRect(0, 0, w, h);
  const strike = strikeProgress(elapsedMs);
  if (strike <= 0) return;

  const minDim = Math.min(w, h);
  const seed = params.boltSeed;
  const jagged = params.jagged;
  const timeS = elapsedMs / 1000;
  const beat = Math.floor(timeS * 12) * params.style;

  // Strike geometry: from near the top edge (biased toward the strike x) down to
  // the strike point (anchor). Canvas y-down: top is y≈0.
  const jx = (hash21(seed * 1.7).x - 0.5) * w * 0.5;
  const A: Vec2 = { x: clamp(center.x + jx, w * 0.12, w * 0.88), y: -0.02 * h };
  const B: Vec2 = { x: center.x, y: center.y };

  ctx.save();
  ctx.globalCompositeOperation = "lighter";

  // MAIN BOLT.
  const radMain = minDim * params.thickness;
  strokeBolt(ctx, boltPolyline(A, B, strike, 0, 1, seed, jagged, beat), radMain);

  // SECONDARY FORKS.
  const forks = Math.max(0, Math.min(MAX_FORKS, Math.round(params.branches)));
  const dlen = Math.hypot(B.x - A.x, B.y - A.y) || 1;
  const dirx = (B.x - A.x) / dlen, diry = (B.y - A.y) / dlen;
  const nrmx = -diry, nrmy = dirx;
  const radFork = minDim * params.thickness * 0.6;
  for (let i = 0; i < forks; i++) {
    const hh = hash21(i * 9.7 + seed + 3);
    const launchT = 0.18 + hh.x * 0.62;
    if (strike < launchT) continue;
    const forkA = boltPoint(A, B, launchT, 0, 1, seed, jagged, beat);
    const ang = (hh.y - 0.5) * 2.2;
    const reach = (0.18 + hh.x * 0.22) * dlen;
    const ex = dirx * (0.5 + hh.y) + nrmx * ang;
    const ey = diry * (0.5 + hh.y) + nrmy * ang;
    const forkB: Vec2 = { x: forkA.x + ex * reach, y: forkA.y + ey * reach };
    const forkDrawn = clamp((strike - launchT) / Math.max(1 - launchT, 0.05), 0, 1);
    strokeBolt(ctx, boltPolyline(forkA, forkB, forkDrawn, i * 17 + 5, 1, seed, jagged, beat), radFork);
  }

  ctx.restore();
}

File: effects/lightning/android/LightningRenderer.kt
// Lightning bolt geometry precompute — faithful port of `lightning-renderer.ts`.
//
// The bolt polyline (trunk + forks) is fragment-INDEPENDENT, so it is computed
// ONCE per frame on the CPU here (a port of the shared fbm/hash + the shader's
// `boltPoint`) and fed to the shader as the `uVerts` / `uBoltMeta` uniform arrays
// (via the backbone's `frameArrays` seam). The shader keeps the exact original
// inverse-distance plasma glow. Output is gl_FragCoord space (device px, y-UP).
//
// Note: these CPU hash/fbm are intentionally the JS port's own (NOT the GLSL
// `Look` chunks) — they reproduce the GLSL noise on the CPU so the precomputed
// polyline matches what the per-pixel shader would have derived.

package ai.dopamine.effect.lightning

import kotlin.math.PI
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.sin

private fun clamp(x: Double, lo: Double, hi: Double): Double = if (x < lo) lo else if (x > hi) hi else x
private fun clamp01(x: Double): Double = if (x < 0.0) 0.0 else if (x > 1.0) 1.0 else x
private fun smoothstep(e0: Double, e1: Double, x: Double): Double {
    val t = clamp01((x - e0) / (e1 - e0))
    return t * t * (3.0 - 2.0 * t)
}
private fun fract(x: Double): Double = x - floor(x)
private fun mix(a: Double, b: Double, t: Double): Double = a + (b - a) * t

// --- Faithful port of the shared look/glsl hash + value-noise fbm -------------
private fun hash11(p0: Double): Double {
    var p = fract(p0 * 0.1031)
    p *= p + 33.33
    p *= p + p
    return fract(p)
}
private fun hash21x(p: Double): Double {
    var x = fract(p * 0.1031); var y = fract(p * 0.103); var z = fract(p * 0.0973)
    val d = x * (y + 33.33) + y * (z + 33.33) + z * (x + 33.33)
    x += d; y += d; z += d
    return fract((x + y) * z)
}
private class V2(val x: Double, val y: Double)
private fun hash21(p: Double): V2 {
    var x = fract(p * 0.1031); var y = fract(p * 0.103); var z = fract(p * 0.0973)
    val d = x * (y + 33.33) + y * (z + 33.33) + z * (x + 33.33)
    x += d; y += d; z += d
    return V2(fract((x + y) * z), fract((x + z) * y))
}
private fun vnoise(x: Double, y: Double): Double {
    val ix = floor(x); val iy = floor(y)
    val fx = x - ix; val fy = y - iy
    val ux = fx * fx * (3.0 - 2.0 * fx); val uy = fy * fy * (3.0 - 2.0 * fy)
    val a = hash11(ix * 1.0 + iy * 57.0)
    val b = hash11((ix + 1.0) * 1.0 + iy * 57.0)
    val c = hash11(ix * 1.0 + (iy + 1.0) * 57.0)
    val d = hash11((ix + 1.0) * 1.0 + (iy + 1.0) * 57.0)
    return mix(mix(a, b, ux), mix(c, d, ux), uy)
}
private fun fbm(x0: Double, y0: Double): Double {
    var x = x0; var y = y0; var s = 0.0; var a = 0.5
    for (i in 0 until 4) {
        s += a * vnoise(x, y)
        val nx = (0.8 * x + 0.6 * y) * 2.03
        val ny = (-0.6 * x + 0.8 * y) * 2.03
        x = nx; y = ny; a *= 0.5
    }
    return s
}

/** Port of the shader boltPoint: a jagged vertex at t along A→B. */
private fun boltPoint(ax: Double, ay: Double, bx: Double, by: Double, t: Double, seedOff: Double, seed: Double, jagged: Double, beat: Double): V2 {
    val dx = bx - ax; val dy = by - ay
    val len = maxOf(hypot(dx, dy), 1.0)
    val dirx = dx / len; val diry = dy / len
    val nrmx = -diry; val nrmy = dirx
    val n = fbm(t * 6.0 + seedOff + seed, beat * 0.5) - 0.5
    val fine = fbm(t * 22.0 + seedOff * 3.1 + seed, beat) - 0.5
    val taper = sin(t * PI)
    val off = (n * 1.6 + fine * 0.5) * jagged * len * 0.16 * taper
    return V2(ax + dirx * (t * len) + nrmx * off, ay + diry * (t * len) + nrmy * off)
}

/** Write up to BOLT_SEGS+1 vertices of the drawn (0..drawn) polyline A→B into
 *  `verts` at bolt slot `b`; returns the segment count (points-1). */
private fun writeBolt(
    verts: FloatArray, b: Int, ax: Double, ay: Double, bx: Double, by: Double, drawn: Double,
    seedOff: Double, seed: Double, jagged: Double, beat: Double,
): Int {
    val base = b * VERTS_PER_BOLT
    var last = 0
    val v0 = boltPoint(ax, ay, bx, by, 0.0, seedOff, seed, jagged, beat)
    verts[(base + 0) * 2] = v0.x.toFloat()
    verts[(base + 0) * 2 + 1] = v0.y.toFloat()
    for (i in 1..BOLT_SEGS) {
        val t = i.toDouble() / BOLT_SEGS
        if (t - 1.0 / BOLT_SEGS > drawn) break
        val tc = minOf(t, drawn)
        val v = boltPoint(ax, ay, bx, by, tc, seedOff, seed, jagged, beat)
        verts[(base + i) * 2] = v.x.toFloat()
        verts[(base + i) * 2 + 1] = v.y.toFloat()
        last = i
    }
    return last
}

/** verts: MAX_BOLTS*VPB*2 ; meta: MAX_BOLTS*4 = (segCount, radFrac, fadeMul, isMain). */
class LightningArrays(val verts: FloatArray, val meta: FloatArray)

/**
 * Compute the bolt polyline (trunk + forks) for this frame, in gl_FragCoord space
 * (device px, y-up). `originX/originY` is the strike point (gl coords).
 */
fun computeLightningArrays(
    style: Double,
    thickness: Double,
    jagged: Double,
    branches: Double,
    boltSeed: Double,
    width: Int,
    height: Int,
    originX: Double,
    originY: Double,
    elapsedMs: Double,
    life: Double,
): LightningArrays {
    val verts = FloatArray(MAX_BOLTS * VERTS_PER_BOLT * 2)
    val meta = FloatArray(MAX_BOLTS * 4)
    val strike = strikeProgress(elapsedMs)
    if (strike <= 0.0) return LightningArrays(verts, meta)

    val w = width.toDouble(); val h = height.toDouble()
    val seed = boltSeed
    val beat = floor((elapsedMs / 1000.0) * 12.0) * style

    // Strike geometry: from near the top edge (biased toward the strike x) down to
    // the strike point. gl coords y-up: top edge is y ≈ height.
    val jx = (hash21x(seed * 1.7) - 0.5) * w * 0.5
    val ax = clamp(originX + jx, w * 0.12, w * 0.88); val ay = h * 1.02
    val bx = originX; val by = originY

    // MAIN BOLT (slot 0).
    val mainSegs = writeBolt(verts, 0, ax, ay, bx, by, strike, 0.0, seed, jagged, beat)
    meta[0] = mainSegs.toFloat(); meta[1] = thickness.toFloat(); meta[2] = 1.0f; meta[3] = 1.0f

    // FORKS (slots 1..).
    val forks = maxOf(0, minOf(MAX_FORKS, Math.round(branches).toInt()))
    val dlen = hypot(bx - ax, by - ay).let { if (it == 0.0) 1.0 else it }
    val dirx = (bx - ax) / dlen; val diry = (by - ay) / dlen
    val nrmx = -diry; val nrmy = dirx
    val forkFade = 0.6 + 0.4 * (1.0 - smoothstep(0.5, 1.0, life))
    for (i in 0 until forks) {
        val b = 1 + i
        val hh = hash21(i * 9.7 + seed + 3.0)
        val launchT = 0.18 + hh.x * 0.62
        if (strike < launchT) { meta[b * 4] = 0f; continue }
        val forkA = boltPoint(ax, ay, bx, by, launchT, 0.0, seed, jagged, beat)
        val ang = (hh.y - 0.5) * 2.2
        val reach = (0.18 + hh.x * 0.22) * dlen
        val ex = dirx * (0.5 + hh.y) + nrmx * ang
        val ey = diry * (0.5 + hh.y) + nrmy * ang
        val forkBx = forkA.x + ex * reach; val forkBy = forkA.y + ey * reach
        val forkDrawn = clamp((strike - launchT) / maxOf(1.0 - launchT, 0.05), 0.0, 1.0)
        val segs = writeBolt(verts, b, forkA.x, forkA.y, forkBx, forkBy, forkDrawn, i * 17.0 + 5.0, seed, jagged, beat)
        meta[b * 4] = segs.toFloat(); meta[b * 4 + 1] = (thickness * 0.6).toFloat()
        meta[b * 4 + 2] = forkFade.toFloat(); meta[b * 4 + 3] = 0f
    }

    return LightningArrays(verts, meta)
}
